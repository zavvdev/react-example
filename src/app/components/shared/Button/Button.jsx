import cx from "clsx";
import PropTypes from "prop-types";
import { childrenPropType } from "app/propTypes/children";
import { useButtonStyles } from "app/components/shared/Button/Button.styles";

export function Button({ children, className, onClick, disabled }) {
  const classes = useButtonStyles();

  const rootClasses = cx(
    classes.root,
    {
      [classes.disabled]: disabled,
    },
    className,
  );

  return (
    <button
      type="button"
      disabled={disabled}
      className={rootClasses}
      onClick={onClick}
    >
      {children}
    </button>
  );
}

Button.propTypes = {
  children: childrenPropType.isRequired,
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  disabled: PropTypes.bool,
};

Button.defaultProps = {
  className: undefined,
  disabled: false,
};
