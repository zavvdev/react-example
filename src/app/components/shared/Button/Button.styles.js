import { createUseStyles } from "react-jss";

export const useButtonStyles = createUseStyles((theme) => ({
  root: {
    backgroundColor: theme.palette.primary.main,
    border: "none",
    color: theme.palette.white.main,
    padding: "5px 20px",
    fontSize: "15px",
    fontWeight: "bold",
    borderRadius: "2px",
    cursor: "pointer",
    userSelect: "none",
    width: "max-content",
    "&:hover": {
      backgroundColor: theme.palette.primary.dark,
    },
  },
  disabled: {
    cursor: "default",
    pointerEvents: "none",
    opacity: 0.8,
  },
}));
