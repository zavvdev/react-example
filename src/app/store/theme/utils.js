import { localStorageService } from "app/services/LocalStorageService";
import { IS_DARK_THEME_LOCAL_STORAGE_KEY } from "app/store/theme/config";

export const saveDarkModeState = (darkModeState) => {
  localStorageService.set(IS_DARK_THEME_LOCAL_STORAGE_KEY, `${darkModeState}`);
};

export const getDarkModeState = () => {
  const rawStorageValue = localStorageService.get(
    IS_DARK_THEME_LOCAL_STORAGE_KEY,
  );
  return !!JSON.parse(rawStorageValue);
};
