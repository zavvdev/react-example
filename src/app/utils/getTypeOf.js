export function getTypeOf(object) {
  const objectType = typeof object;
  let returnType = objectType;
  if (objectType === "object") {
    const stringifiedObject = Object.prototype.toString.call(object);
    returnType = stringifiedObject
      .replace(/^\[object (\S+)\]$/, "$1")
      .toLowerCase();
  }
  return returnType;
}
