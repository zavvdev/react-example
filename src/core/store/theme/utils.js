import { LOCAL_STORAGE_KEYS } from "core/config/localStorage";
import { localStorageService } from "core/services/LocalStorageService";

export const saveDarkModeState = (darkModeState) => {
  localStorageService.set(
    LOCAL_STORAGE_KEYS.isThemeDarkMode,
    `${darkModeState}`,
  );
};

export const getDarkModeState = () => {
  const rawStorageValue = localStorageService.get(
    LOCAL_STORAGE_KEYS.isThemeDarkMode,
  );
  return !!JSON.parse(rawStorageValue);
};
