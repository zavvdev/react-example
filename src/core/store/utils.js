import { all, call, spawn } from "redux-saga/effects";
import { errorTrackingService } from "core/services/ErrorTrackingService";

export const combineMiddleware = (middlewareList) => function* combiner() {
  const getRootGenerator = (middleware) => function* rootGenerator() {
    while (true) {
      try {
        yield call(middleware);
        break;
      } catch (e) {
        errorTrackingService.reportError(e);
      }
    }
  };

  yield all(middlewareList.map((middleware) => {
    return spawn(getRootGenerator(middleware));
  }));
};
