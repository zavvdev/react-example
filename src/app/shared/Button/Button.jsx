import cx from "clsx";
import PropTypes from "prop-types";
import { childrenPropType } from "app/propTypes/children";
import { useButtonStyles } from "app/shared/Button/Button.styles";

export function Button({ children, className, onClick, disabled, fullWidth }) {
  const classes = useButtonStyles();

  const rootClasses = cx(
    classes.root,
    {
      [classes.disabled]: disabled,
      [classes.rootFullWidth]: fullWidth,
    },
    className,
  );

  return (
    <button
      type="button"
      disabled={disabled}
      className={rootClasses}
      onClick={onClick}
    >
      {children}
    </button>
  );
}

Button.propTypes = {
  children: childrenPropType.isRequired,
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  fullWidth: PropTypes.bool,
};

Button.defaultProps = {
  className: undefined,
  disabled: false,
  fullWidth: false,
};
