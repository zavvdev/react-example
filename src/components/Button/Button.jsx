import cx from "clsx";
import PropTypes from "prop-types";
import { useButtonStyles } from "components/Button/Button.styles";
import { childrenPropType } from "propTypes/children";

export function Button({
  children,
  className,
  onClick,
  disabled,
  icon,
}) {
  const classes = useButtonStyles();

  const rootClasses = cx(classes.root, {
    [classes.disabled]: disabled,
  }, className);

  return (
    <button
      type="button"
      disabled={disabled}
      className={rootClasses}
      onClick={onClick}
    >
      {icon && <div className={classes.icon}>{icon}</div>}
      {children}
    </button>
  );
}

Button.propTypes = {
  children: childrenPropType.isRequired,
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  icon: PropTypes.node,
};

Button.defaultProps = {
  className: undefined,
  disabled: false,
  icon: undefined,
};
