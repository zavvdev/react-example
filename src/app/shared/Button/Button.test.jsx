import { fireEvent, screen } from "@testing-library/react";
import { testUtils } from "app/tests/utils";
import { Button } from "app/shared/Button/Button";

const properties = {
  children: "Button",
  className: "classNameMock",
  onClick: jest.fn(),
};

describe("Button", () => {
  test("should render Button", () => {
    const result = testUtils.render(
      <Button onClick={properties.onClick}>{properties.children}</Button>,
    );
    expect(result.asFragment()).toMatchSnapshot();
  });

  test("should have a custom className", () => {
    const result = testUtils.render(
      <Button onClick={properties.onClick} className={properties.className}>
        {properties.children}
      </Button>,
    );
    expect(
      result.container.querySelectorAll(`.${properties.className}`).length,
    ).toBe(1);
  });

  test("should have a disabled className", () => {
    const result = testUtils.render(
      <Button onClick={properties.onClick} disabled>
        {properties.children}
      </Button>,
    );
    expect(result.container.querySelectorAll(".disabled").length).toBe(1);
  });

  test("should have a full width className", () => {
    const result = testUtils.render(
      <Button onClick={properties.onClick} fullWidth>
        {properties.children}
      </Button>,
    );
    expect(result.container.querySelectorAll(".rootFullWidth").length).toBe(1);
  });

  test("should call onClick function", () => {
    testUtils.render(
      <Button onClick={properties.onClick}>{properties.children}</Button>,
    );
    fireEvent.click(screen.getByText(properties.children));
    expect(properties.onClick).toBeCalled();
  });
});
