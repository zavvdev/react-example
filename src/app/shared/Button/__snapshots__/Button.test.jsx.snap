// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render Button 1`] = `
<DocumentFragment>
  <button
    class="root"
    type="button"
  >
    Button
  </button>
</DocumentFragment>
`;
